pkg/ebpf/events_enrich.go

```diff
@@ -8,172 +8,157 @@ import (
+	enrichDone := make(map[uint64]*uberAtom.Bool)
+	enrichInfo := make(map[uint64]*enrichResult)
+	// 1 queue per cgroupId
+	queues := make(map[uint64]chan *trace.Event)
+	// scheduler queue
+	queueReady := make(chan uint64, queueReadySize)
+
+	// queues map writer
 	go func() {
 		defer close(out)
 		defer close(errc)
-		for {
+		for { // enqueue events
 			select {
-			case <-ctx.Done():
-				done <- struct{}{}
-				return
 			case event := <-in:
+				// send out irrelevant events
+				if event.ContainerID == "" {
+					out <- event
+					continue
+				}
 				cgroupId := uint64(event.CgroupID)
-				// cgroup_mkdir event: need the cgroupId from its argument
+				// CgroupMkdir: pick EventID from the event itself
 				if event.EventID == int(events.CgroupMkdir) {
 					cgroupId, _ = parse.ArgUint64Val(event, "cgroup_id")
 				}
-				// cgroup_rmdir: need to clean attempt so cgroupId can be reused
+				// CgroupRmdir: clean up remaining events and maps
 				if event.EventID == int(events.CgroupRmdir) {
 					cgroupId, _ = parse.ArgUint64Val(event, "cgroup_id")
-					attemptedMutex.Lock()
-					delete(attempted, cgroupId)
-					attemptedMutex.Unlock()
-				}
-				// non container event and not cgroup_mkdir (or the event already enriched) skip per cgroupId caching
-				if (event.ContainerID == "" && event.EventID != int(events.CgroupMkdir)) || event.ContainerImage != "" {
-					// enrichment attempt is being done, or has failed, so do not try it again
-					attemptedMutex.RLock()
-					if _, ok := attempted[cgroupId]; ok {
-						// wait transaction (cache flush) to finish so cached events and this one are ordered
-						attempted[cgroupId].Lock()
-						out <- event
-						attempted[cgroupId].Unlock()
-						attemptedMutex.RUnlock()
-						continue
-					}
-					attemptedMutex.RUnlock()
-					out <- event // some events might not have an attempted map
-					continue
-				} else {
-					//if the container queue has not been created - create the container queue and invoke to enrich query
-					queuesMutex.RLock()
-					_, exists := queues[cgroupId]
-					queuesMutex.RUnlock()
-					if !exists {
-						queuesMutex.Lock()
-						queues[cgroupId] = make(chan *trace.Event, enrichmentQueueSize)
-						queuesMutex.Unlock()
-
-						go func() {
-							metadata, err := t.containers.EnrichCgroupInfo(cgroupId)
-							enriches <- enrichResult{cgroupId, metadata, err}
-						}()
-					}
-					queuesMutex.RLock()
-					queue, ok := queues[cgroupId]
-					queuesMutex.RUnlock()
-					if ok {
-						//if queue still exists send as usual
-						queue <- event
-					} else {
-						//if not, get the enriched data and handle the stale queue
-						info := t.containers.GetCgroupInfo(cgroupId)
-						queuesMutex.RLock()
-						queue = staleQueues[cgroupId]
-						queuesMutex.RUnlock()
-
-						//probably empty but just to make sure
-						for evt := range queue {
-							enrichEvent(evt, info.Container)
-							out <- evt
+					i := enrichInfo[cgroupId]
```

Comment:
what if `enrichInfo[cgroupId]` entry doesn't exist?
This might happen, for example, when starting tracee with pre-existing containers which are then removed

One more point - the (read) lock is not taken for this operation - shouldn't it be?

Reply:
Yep, added a read lock and addressed the map key existence.